readthedocs/builds/models.py

```diff
@@ -461,11 +498,68 @@ class Build(models.Model):
 
     objects = BuildQuerySet.as_manager()
 
+    CONFIG_KEY = '__config'
+
     class Meta(object):
         ordering = ['-date']
         get_latest_by = 'date'
         index_together = [['version', 'state', 'type']]
 
+    @property
+    def previous(self):
+        """
+        Returns the previous build to current one.
+
+        Matching the project and version.
+        """
+        date = self.date or timezone.now()
+        if self.project is not None and self.version is not None:
+            return (
+                Build.objects
+                .filter(
+                    project=self.project,
+                    version=self.version,
+                    date__lt=date,
+                )
+                .order_by('-date')
+                .first()
+            )
+        return None
+
+    @property
+    def config(self):
+        """
+        Get the config used for this build.
+
+        Since we are saving the config into the JSON field only when it differs
+        from the previous one, this helper returns the correct JSON used in
+        this Build object (it could be stored in this object or one of the
+        previous ones).
+        """
+        if self.CONFIG_KEY in self._config:
+            return Build.objects.get(pk=self._config[self.CONFIG_KEY])._config
+        return self._config
+
+    @config.setter
+    def config(self, value):
+        self._config = value
+
+    def save(self, *args, **kwargs):  # noqa
+        """
+        Save object.
+
+        To save space on the db we only save the config if it's different
+        from the previous one.
+
+        If the config is the same, we save the pk of the object
+        that has the **real** config under the `CONFIG_KEY` key.
+        """
+        previous = self.previous
+        if previous is not None and self._config == previous.config:
+            previous_pk = previous._config.get(self.CONFIG_KEY, previous.pk)
+            self._config = {self.CONFIG_KEY: previous_pk}
+        super(Build, self).save(*args, **kwargs)
```

Comment:
I'm a little worried that we're running this on each `Build` save. We do this a number of times during a documentation build, and this will be adding a decent number of queries (eg. `self.previous` is a query, as is `previous.config`).

Is there a way to set the `_config` just once when we have changed it?

Reply:
I don't think there is an `is_changed` method on django, but I think we can do this using an internal variable.

Reply:
Yea, I'm just trying to think if there is a way to run this logic _just_ when we are trying to set a `config`. Perhaps it belongs somewhere else?